ecosystem/gstreamer_plugin/gst_mtl_st40_rx.c

```diff
@@ -0,0 +1,509 @@
+    return;
+  }
+
+  switch (prop_id) {
+    case PROP_ST40_RX_BUFFER_SIZE:
+      g_value_set_uint(value, src->mbuff_size);
+      break;
+    case PROP_ST40_RX_TIMEOUT_MBUF_GET:
+      g_value_set_uint(value, src->timeout_mbuf_get_seconds);
+      break;
+    default:
+      G_OBJECT_WARN_INVALID_PROPERTY_ID(object, prop_id, pspec);
+      break;
+  }
+}
+
+static void* gst_mtl_st40_rx_get_mbuf_with_timeout(Gst_Mtl_St40_Rx* src,
+                                                   st40_rx_handle handle, void** usrptr,
+                                                   uint16_t* size) {
+  struct timespec ts;
+  gint ret;
+  void* mbuf = NULL;
+
+  clock_gettime(CLOCK_REALTIME, &ts);
+  ts.tv_sec += src->timeout_mbuf_get_seconds;
+  ret = pthread_mutex_timedlock(&(src->get_mbuff_mutex), &ts);
+
+  if (!ret)
+    ret = pthread_cond_timedwait(&(src->get_mbuff_cond), &(src->get_mbuff_mutex), &ts);
+
+  if (ret == ETIMEDOUT) return NULL;
+
+  pthread_mutex_unlock(&(src->get_mbuff_mutex));
+
+  mbuf = st40_rx_get_mbuf(src->rx_handle, usrptr, size);
+  if (!mbuf) GST_ERROR("Failed to get ancillary mbuf\n");
+
+  return mbuf;
+}
+
+static GstFlowReturn gst_mtl_st40_rx_fill_buffer(Gst_Mtl_St40_Rx* src, GstBuffer** buffer,
+                                                 void* usrptr) {
+  struct st40_rfc8331_rtp_hdr* hdr;
+  struct st40_rfc8331_payload_hdr* payload_hdr;
+  GstMapInfo dest_info;
+  guint16 data, fill_size;
+  gint udw_size;
+
+  hdr = (struct st40_rfc8331_rtp_hdr*)usrptr;
+  payload_hdr = (struct st40_rfc8331_payload_hdr*)(&hdr[1]);
+  udw_size = payload_hdr->second_hdr_chunk.data_count & 0xff;
+
+  if (src->udw_size == 0) {
+    src->udw_size = udw_size;
+    src->anc_data = (char*)malloc(udw_size);
+  } else if (src->udw_size != udw_size) {
+    GST_WARNING("Size of recieved ancillary data has changed");
+    if (src->anc_data) {
+      free(src->anc_data);
+      src->anc_data = NULL;
+    }
+    src->udw_size = udw_size;
+    src->anc_data = (char*)malloc(udw_size);
+  }
+
+  if (udw_size == 0) {
+    GST_ERROR("Ancillary data size is 0");
+    return GST_FLOW_ERROR;
+  }
+
+  *buffer = gst_buffer_new_allocate(NULL, src->udw_size, NULL);
+  if (!*buffer) {
+    GST_ERROR("Failed to allocate space for the buffer");
+    return GST_FLOW_ERROR;
+  }
+
+  if (!gst_buffer_map(*buffer, &dest_info, GST_MAP_WRITE)) {
+    GST_ERROR("Failed to map the buffer");
+    return GST_FLOW_ERROR;
+  }
+
+  for (int i = 0; i < udw_size; i++) {
+    data = st40_get_udw(i + 3, (uint8_t*)&payload_hdr->second_hdr_chunk);
+    if (!st40_check_parity_bits(data)) {
+      GST_ERROR("Ancillary data parity bits check failed");
+      return GST_FLOW_ERROR;
+    }
+    src->anc_data[i] = data & 0xff;
+  }
+
+  fill_size = gst_buffer_fill(*buffer, 0, src->anc_data, udw_size);
+  gst_buffer_unmap(*buffer, &dest_info);
+
+  if (fill_size != src->udw_size) {
+    GST_WARNING("Failed to fill buffer");
```

Comment:
do we want to return ok anyway?

Reply:
well i kinda think we can continue as long as we warn here
hmmm 
I will change this to gst-error lest stay on the safe side